docs/git-workflow.md

```diff
@@ -25,9 +25,25 @@ The Jetpack repo uses the following "reserved" branch name conventions:
 
 ## Keeping Your Branch Up To Date
 
-While it is tempting to merge from `master` into your branch frequently, this leads to a messy history because each merge creates a merge commit. When working by yourself, it is best to use `git pull --rebase master`, but if you're pushing to a shared repo, it is best to not do any merging or rebasing until the feature is ready for final testing, and then do a [rebase](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request) at the very end. This is one reason why it is important to open pull requests whenever you have working code.
-
-If you have a Pull Request branch that cannot be merged into `master` due to a conflict (this can happen for long-running Pull Request discussions), it's still best to rebase the branch (rather than merge) and resolve any conflicts on your local copy before updating the Pull Request with `git push --force`. **Be aware** that this will **replace** any commits currently in your shared branch, so anyone who is also using that branch will be in trouble. Only use `git push --force` if the Pull Request is ready to merge and no one else is using it (or of you have coordinated the force-push with the other developers working on the branch).
+There are two ways to update your branch with changes to `master` or a parent PR.
+
+1. **Merge:** After a `git fetch`, execute `git merge origin/master` to pull in the latest changes. If there are conflicts, you'll need to resolve them by hand and then commit manually.
+   * Pro: GitHub deals with this much better.
+   * Pro: Collaborators can just `git pull` to update their local copy of the branch.
+   * Con: Can lead to a messy-looking history while the branch is in progress. Since we use "squash and merge" to commit PRs, though, this isn't an issue once the PR is finally accepted.
+   * Con: If there's a conflict while merging, you might wind up having to commit the merge commit with `--no-verify` to avoid our pre-commit hook complaining about things.
+   * Note: To get a clean view of the branch's history, use `git log --first-parent`.
+   * Note: To get a clean diff of the branch versus master, like what GitHub shows, use `git diff origin/master...HEAD` (note the three dots).
+2. **[Rebase](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request):** Execute `git pull --rebase origin master`, or do a `git fetch` then execute `git rebase origin/master`. It will reapply each patch, and if there are conflicts at any step you'll have to resolve them by hand and then `git rebase --continue`.
+   * Pro: Keeps the branch's history cleaner.
+   * Con: GitHub doesn't handle it very well. It may lose inline comments on the pre-rebase commits, and it will show every commit in the rebase as being "added" again which clutters the GitHub conversation page.
```

Comment:
I believe the old commits get removed and replaced by the new, rebased commits so that shouldn't add much to the conversation page.

[Here is an example](https://github.com/Automattic/jetpack/pull/18054).

The first part of the statement, however, can definitely be a problem. 👍

```suggestion
   * Con: GitHub doesn't handle it very well. It may lose inline comments on the pre-rebase commits.
```

Reply:
Hmm, that seems arguably even worse, as whatever old comments do stay can't easily be matched up with the state of the PR since all the commits get moved down where they'll clutter the active part of the discussion.